# ---------------------------------------------------
# Advanced Apex Project: Real Estate Price Prediction
# Team: The Outliers
# ---------------------------------------------------

## Phase 2 : Preprocessing and Feature Engineering

In [10]:
# STEP 2: Import required libraries
import pandas as pd
import numpy as np

In [11]:
# STEP 3: Load dataset
file_path = "./data/AmesHousing.csv"   # ensure this file exists in your project folder
df = pd.read_csv(file_path)

In [12]:
# STEP 4: Basic inspection
print("Dataset shape:", df.shape)
print("\nColumns:", df.columns.tolist()[:10], "...")
print("\nFirst 5 rows:")
display(df.head())



Dataset shape: (2930, 82)

Columns: ['Order', 'PID', 'MS SubClass', 'MS Zoning', 'Lot Frontage', 'Lot Area', 'Street', 'Alley', 'Lot Shape', 'Land Contour'] ...

First 5 rows:


,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,...,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
0,1,526301100,20,RL,141.0,31770,Pave,NaN,IR1,Lvl,...,0,NaN,NaN,NaN,0,5,2010,WD,Normal,215000
1,2,526350040,20,RH,80.0,11622,Pave,NaN,Reg,Lvl,...,0,NaN,MnPrv,NaN,0,6,2010,WD,Normal,105000
2,3,526351010,20,RL,81.0,14267,Pave,NaN,IR1,Lvl,...,0,NaN,NaN,Gar2,12500,6,2010,WD,Normal,172000
3,4,526353030,20,RL,93.0,11160,Pave,NaN,Reg,Lvl,...,0,NaN,NaN,NaN,0,4,2010,WD,Normal,244000
4,5,527105010,60,RL,74.0,13830,Pave,NaN,IR1,Lvl,...,0,NaN,MnPrv,NaN,0,3,2010,WD,Normal,189900


In [13]:
# STEP 5: Data types and non-null counts
print("\nData Types and Non-Null Counts:")
print(df.info())




Data Types and Non-Null Counts:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2930 non-null   int64  
 1   PID              2930 non-null   int64  
 2   MS SubClass      2930 non-null   int64  
 3   MS Zoning        2930 non-null   object 
 4   Lot Frontage     2440 non-null   float64
 5   Lot Area         2930 non-null   int64  
 6   Street           2930 non-null   object 
 7   Alley            198 non-null    object 
 8   Lot Shape        2930 non-null   object 
 9   Land Contour     2930 non-null   object 
 10  Utilities        2930 non-null   object 
 11  Lot Config       2930 non-null   object 
 12  Land Slope       2930 non-null   object 
 13  Neighborhood     2930 non-null   object 
 14  Condition 1      2930 non-null   object 
 15  Condition 2      2930 non-null   object 
 16  Bldg Type        2930 non-n

In [14]:
# STEP 6: Quick statistics (numeric columns)
print("\nSummary statistics (numeric features):")
display(df.describe().T.head(10))



Summary statistics (numeric features):


,count,mean,std,min,25%,50%,75%,max
Order,2930.0,1.465500e+03,8.459625e+02,1.0,7.332500e+02,1465.5,2.197750e+03,2.930000e+03
PID,2930.0,7.144645e+08,1.887308e+08,526301100.0,5.284770e+08,535453620.0,9.071811e+08,1.007100e+09
MS SubClass,2930.0,5.738737e+01,4.263802e+01,20.0,2.000000e+01,50.0,7.000000e+01,1.900000e+02
Lot Frontage,2440.0,6.922459e+01,2.336533e+01,21.0,5.800000e+01,68.0,8.000000e+01,3.130000e+02
Lot Area,2930.0,1.014792e+04,7.880018e+03,1300.0,7.440250e+03,9436.5,1.155525e+04,2.152450e+05
Overall Qual,2930.0,6.094881e+00,1.411026e+00,1.0,5.000000e+00,6.0,7.000000e+00,1.000000e+01
Overall Cond,2930.0,5.563140e+00,1.111537e+00,1.0,5.000000e+00,5.0,6.000000e+00,9.000000e+00
Year Built,2930.0,1.971356e+03,3.024536e+01,1872.0,1.954000e+03,1973.0,2.001000e+03,2.010000e+03
Year Remod/Add,2930.0,1.984267e+03,2.086029e+01,1950.0,1.965000e+03,1993.0,2.004000e+03,2.010000e+03
Mas Vnr Area,2907.0,1.018968e+02,1.791126e+02,0.0,0.000000e+00,0.0,1.640000e+02,1.600000e+03


In [15]:
# STEP 7: Missing values overview
print("\nMissing values (top 15 features with NaNs):")
print(df.isnull().sum().sort_values(ascending=False).head(15))



Missing values (top 15 features with NaNs):
Pool QC           2917
Misc Feature      2824
Alley             2732
Fence             2358
Mas Vnr Type      1775
Fireplace Qu      1422
Lot Frontage       490
Garage Qual        159
Garage Yr Blt      159
Garage Cond        159
Garage Finish      159
Garage Type        157
Bsmt Exposure       83
BsmtFin Type 2      81
Bsmt Qual           80
dtype: int64


In [16]:
# Histogram of Target Variable (SalePrice)

---

## Closing Notes

- The Ames Housing dataset has been **successfully extracted** using the direct download method from Kaggle and loaded into a Pandas DataFrame.  
- Initial inspection steps (`.shape`, `.head()`, `.info()`, `.describe()`, and missing values summary) confirm that the dataset contains **2,930 rows and 82 columns**, with several features having missing values (e.g., `PoolQC`, `Alley`, `Fence`, `FireplaceQu`).  
- A **concise Schema/Data Dictionary** covering 20 important features has been prepared separately in `data_dictionary_phase1.xlsx`.  
- The **complete data dictionary** (all 82 features) will be maintained and expanded in later phases of the project.  

**Phase 1 Deliverables Completed:**  
1. Proposal Document (PDF/LaTeX)  
2. Data Extraction Notebook (`data_extraction.ipynb`)  
3. Concise Data Dictionary (`data_dictionary_phase1.xlsx`)  

---


### `.describe()`
The `.describe()` method provides descriptive statistics for all numerical columns, including count, mean, standard deviation, min/max, and quartiles.

In [17]:
# Generate descriptive statistics for numerical columns
df.describe()

,Order,PID,MS SubClass,Lot Frontage,Lot Area,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Mas Vnr Area,...,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Misc Val,Mo Sold,Yr Sold,SalePrice
count,2930.00000,2.930000e+03,2930.000000,2440.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2907.000000,...,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000,2930.000000
mean,1465.50000,7.144645e+08,57.387372,69.224590,10147.921843,6.094881,5.563140,1971.356314,1984.266553,101.896801,...,93.751877,47.533447,23.011604,2.592491,16.002048,2.243345,50.635154,6.216041,2007.790444,180796.060068
std,845.96247,1.887308e+08,42.638025,23.365335,7880.017759,1.411026,1.111537,30.245361,20.860286,179.112611,...,126.361562,67.483400,64.139059,25.141331,56.087370,35.597181,566.344288,2.714492,1.316613,79886.692357
min,1.00000,5.263011e+08,20.000000,21.000000,1300.000000,1.000000,1.000000,1872.000000,1950.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,2006.000000,12789.000000
25%,733.25000,5.284770e+08,20.000000,58.000000,7440.250000,5.000000,5.000000,1954.000000,1965.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,4.000000,2007.000000,129500.000000
50%,1465.50000,5.354536e+08,50.000000,68.000000,9436.500000,6.000000,5.000000,1973.000000,1993.000000,0.000000,...,0.000000,27.000000,0.000000,0.000000,0.000000,0.000000,0.000000,6.000000,2008.000000,160000.000000
75%,2197.75000,9.071811e+08,70.000000,80.000000,11555.250000,7.000000,6.000000,2001.000000,2004.000000,164.000000,...,168.000000,70.000000,0.000000,0.000000,0.000000,0.000000,0.000000,8.000000,2009.000000,213500.000000
max,2930.00000,1.007100e+09,190.000000,313.000000,215245.000000,10.000000,9.000000,2010.000000,2010.000000,1600.000000,...,1424.000000,742.000000,1012.000000,508.000000,576.000000,800.000000,17000.000000,12.000000,2010.000000,755000.000000


### `.quantile()`
If you need specific quantiles (percentiles) beyond the ones provided by `.describe()`, you can use the `.quantile()` method.

In [18]:
# Calculate the 10th, 50th (median), and 90th quantiles for 'SalePrice'
price_quantiles = df['SalePrice'].quantile([0.1, 0.5, 0.9])

print("SalePrice Quantiles:")
print(price_quantiles)

SalePrice Quantiles:
0.1    105450.0
0.5    160000.0
0.9    281241.7
Name: SalePrice, dtype: float64


## 3. Categorical Data Analysis with `.value_counts()`

To understand the distribution of categorical variables, `.value_counts()` is extremely useful. It returns a series containing counts of unique values.

In [19]:
# Get the number of houses in each neighborhood
neighborhood_counts = df['Neighborhood'].value_counts()

print("Number of houses per neighborhood:")
print(neighborhood_counts.head(10)) # Display top 10

Number of houses per neighborhood:
Neighborhood
NAmes      443
CollgCr    267
OldTown    239
Edwards    194
Somerst    182
NridgHt    166
Gilbert    165
Sawyer     151
NWAmes     131
SawyerW    125
Name: count, dtype: int64


## 4. Correlation Analysis

Correlation measures the linear relationship between two variables. It ranges from -1 (perfect negative correlation) to +1 (perfect positive correlation). A value near 0 indicates no linear correlation. This is crucial for feature selection in a prediction model.

In [20]:
# Calculate the correlation matrix for all numerical columns
# The 'numeric_only=True' argument is important to avoid errors with non-numeric columns
correlation_matrix = df.corr(numeric_only=True)

# To make this more readable, let's see which features correlate most strongly with 'SalePrice'
saleprice_correlation = correlation_matrix['SalePrice'].sort_values(ascending=False)

print("Top 10 features most positively correlated with SalePrice:")
print(saleprice_correlation.head(11)) # a variable is always 100% correlated with itself

print("\nTop 5 features most negatively correlated with SalePrice:")
print(saleprice_correlation.tail(5))

Top 10 features most positively correlated with SalePrice:
SalePrice         1.000000
Overall Qual      0.799262
Gr Liv Area       0.706780
Garage Cars       0.647877
Garage Area       0.640401
Total Bsmt SF     0.632280
1st Flr SF        0.621676
Year Built        0.558426
Full Bath         0.545604
Year Remod/Add    0.532974
Garage Yr Blt     0.526965
Name: SalePrice, dtype: float64

Top 5 features most negatively correlated with SalePrice:
MS SubClass      -0.085092
Overall Cond     -0.101697
Kitchen AbvGr    -0.119814
Enclosed Porch   -0.128787
PID              -0.246521
Name: SalePrice, dtype: float64


## 5. Slicing and Dicing (Data Subsetting)

Slicing and dicing involves selecting specific rows and columns to isolate a subset of the data for analysis. We typically use `.loc[]` for label-based indexing and `.iloc[]` for position-based indexing.

### Column Selection

In [21]:
# Select a few specific columns to create a smaller DataFrame
subset_df = df[['Lot Area', 'Neighborhood', 'Year Built', 'Overall Qual', 'SalePrice']]
subset_df.head()

,Lot Area,Neighborhood,Year Built,Overall Qual,SalePrice
0,31770,NAmes,1960,6,215000
1,11622,NAmes,1961,5,105000
2,14267,NAmes,1958,6,172000
3,11160,NAmes,1968,7,244000
4,13830,Gilbert,1997,5,189900


### Row Selection with Conditions using `.loc`

In [22]:
# Select all houses built after the year 2000
modern_houses = df.loc[df['Year Built'] > 2000]

print(f"Number of houses built after 2000: {len(modern_houses)}")
modern_houses[['Year Built', 'SalePrice']].head()

Number of houses built after 2000: 735


,Year Built,SalePrice
6,2001,213500
15,2003,538000
17,2010,394432
36,2009,376162
37,2007,306000


In [23]:
# You can also combine multiple conditions.
# Let's find houses with an 'Overall Qual' of 8 or more that sold for less than $200,000
high_quality_bargains = df.loc[(df['Overall Qual'] >= 8) & (df['SalePrice'] < 200000)]

print(f"Number of high-quality bargain houses: {len(high_quality_bargains)}")
high_quality_bargains[['Overall Qual', 'SalePrice', 'Neighborhood']].head()

Number of high-quality bargain houses: 40


,Overall Qual,SalePrice,Neighborhood
7,8,191500,StoneBr
16,8,164000,Gilbert
53,8,192000,Blmngtn
106,8,155000,Greens
183,8,150000,OldTown


## 6. Roll-up (Grouping and Aggregation)

"Rolling up" data usually means grouping it by some category and then calculating an aggregate statistic for each group (like a sum, mean, or count). This is done with the `.groupby()` method.

In [24]:
# Let's find the average SalePrice for each neighborhood
avg_price_by_neighborhood = df.groupby('Neighborhood')['SalePrice'].mean().sort_values(ascending=False).round(2)

print("Average Sale Price by Neighborhood:")
print(avg_price_by_neighborhood.head(10))

Average Sale Price by Neighborhood:
Neighborhood
NoRidge    330319.13
StoneBr    324229.20
NridgHt    322018.27
GrnHill    280000.00
Veenker    248314.58
Timber     246599.54
Somerst    229707.32
ClearCr    208662.09
Crawfor    207550.83
CollgCr    201803.43
Name: SalePrice, dtype: float64


In [25]:
# You can also perform multiple aggregations at once using .agg()
# Let's group by 'Overall Qual' and find the mean, median, standard deviation, and count of 'SalePrice' for each quality level.

quality_rollup = df.groupby('Overall Qual')['SalePrice'].agg(['mean', 'median', 'count', 'std']).round(2)

print("SalePrice Statistics by Overall Quality:")
print(quality_rollup)

SalePrice Statistics by Overall Quality:
                   mean    median  count        std
Overall Qual                                       
1              48725.00   50150.0      4   29341.94
2              52325.31   57625.0     13   17562.96
3              83185.98   81200.0     40   23569.80
4             106485.10  105000.0    226   29224.94
5             134752.52  133000.0    825   27690.60
6             162130.32  159500.0    732   37201.30
7             205025.76  200000.0    602   43166.27
8             270913.59  264530.5    350   61326.21
9             368336.77  360000.0    107   79201.27
10            450217.32  451950.0     31  141975.97
